optuna/_hypervolume/box_decomposition.py

```diff
@@ -0,0 +1,150 @@
+        paper.
+        ``def_points`` (the shape is ``(n_bounds, n_objectives, n_objectives)``) is not well
+        explained in the paper, but basically, ``def_points[i, j] = z[j]`` of
+        ``upper_bound_set[i]``.
+    """
+    (_, n_objectives) = sorted_pareto_sols.shape
+    objective_indices = np.arange(n_objectives)
+    skip_ineq_judge = np.eye(n_objectives, dtype=bool)
+    # NOTE(nabenabe): True at 0 comes from Line 2 of Alg. 2. (loss_vals is sorted w.r.t. 0-th obj)
+    skip_ineq_judge[:, 0] = True
+
+    def update(sol: np.ndarray, ubs: np.ndarray, dps: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
+        # The update rule is written in Section 2.2 of Lacour17.
+        is_dominated = np.all(sol < ubs, axis=-1)
+        if not any(is_dominated):
+            return ubs, dps
+
+        dominated_dps = dps[is_dominated]  # The set `A` in Line 5 of Alg. 2.
+        n_bounds = dominated_dps.shape[0]
+        # NOTE(nabenabe): `-inf` comes from Line 2 and `k!=j` in Line 3 of Alg. 2.
+        # NOTE(nabenabe): If `update[i,j]=True`, update `ubs[i,j]` to `(z_j, u_{-j})`, i.e.,
+        # `np.where(objective_indices != j, ubs[i], sol[j])` in `new_ubs`. cf. Lines 2,3 of Alg. 2.
+        update = sol >= np.max(np.where(skip_ineq_judge, -np.inf, dominated_dps), axis=-2)
+        # NOTE(nabenabe): The indices of `u` with `True` in update. Each `u` may yield `True`
+        # multiple times for different indices `j`.
+        ubs_indices_to_update = np.tile(np.arange(n_bounds)[:, np.newaxis], n_objectives)[update]
+        # The dimension `j` for each `u` s.t. `\hat{z}_j \geq \max_{k \neq j}{z_j^k(u)}`.
+        dimensions_to_update = np.tile(objective_indices, (n_bounds, 1))[update]
+        assert ubs_indices_to_update.size == dimensions_to_update.size
+        indices_for_sweeping = np.arange(dimensions_to_update.size)
+        # The last Eq in Page 5.
+        new_dps = dominated_dps[ubs_indices_to_update]
+        new_dps[indices_for_sweeping, dimensions_to_update] = sol
+        # Line 3 of Alg. 2. `sol[dimensions_to_update]` is equivalent to `\bar{z}_j`.
+        new_ubs = ubs[is_dominated][ubs_indices_to_update]
+        new_ubs[indices_for_sweeping, dimensions_to_update] = sol[dimensions_to_update]
+        return np.vstack([ubs[~is_dominated], new_ubs]), np.vstack([dps[~is_dominated], new_dps])
+
+    upper_bound_set = np.asarray([ref_point])  # Line 1 of Alg. 2.
+    def_points = np.full((1, n_objectives, n_objectives), -np.inf)  # z^k(z^r) = \hat{z}^k
+    def_points[0, objective_indices, objective_indices] = ref_point  # \hat{z}^k is a dummy point.
+    for solution in sorted_pareto_sols:  # NOTE(nabenabe): Sorted must be fulfilled.
+        upper_bound_set, def_points = update(solution, upper_bound_set, def_points)
+
+    return upper_bound_set, def_points
+
+
+def _get_box_bounds(
+    def_points: np.ndarray, upper_bound_set: np.ndarray, ref_point: np.ndarray
```

Comment:
```suggestion
    upper_bound_set: np.ndarray, def_points: np.ndarray, ref_point: np.ndarray
```
In other parts such as the `update` function, the argument `upper_bound_set` comes before `def_points`, so this ordering is more consistent.